pkg/utils/fetcher/fetchEtcd.go

```diff
@@ -0,0 +1,85 @@
+package fetcher
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/pingcap/log"
+
+	"github.com/pingcap-incubator/tidb-dashboard/pkg/utils/clusterinfo"
+)
+
+const prefix = "/topology"
+
+func FetchEtcd(ctx context.Context, etcdcli *clientv3.Client) ([]clusterinfo.TiDB, clusterinfo.Grafana,
+	clusterinfo.AlertManager, error) {
+	resp, err := etcdcli.Get(ctx, prefix, clientv3.WithPrefix())
+	if err != nil {
+		// put error in ctx and return
+		return nil, clusterinfo.Grafana{}, clusterinfo.AlertManager{}, err
+	}
+	dbMap := make(map[string]*clusterinfo.TiDB)
+	var grafana clusterinfo.Grafana
+	var alertManager clusterinfo.AlertManager
+	dbList := make([]clusterinfo.TiDB, 0)
+	for _, kvs := range resp.Kvs {
+		key := string(kvs.Key)
+		keyParts := strings.Split(key, "/")
+		if keyParts[0] == "" {
+			log.Info("Sure, it is.")
+			keyParts = keyParts[1:]
+		}
+		if len(keyParts) < 2 {
+			continue
+		}
+		log.Info(keyParts[1] + "  " + string(kvs.Value))
+		switch keyParts[1] {
+		case "grafana":
+			if err = json.Unmarshal(kvs.Value, &grafana); err != nil {
+				return nil, clusterinfo.Grafana{}, clusterinfo.AlertManager{}, err
```

Comment:
Why failure in deserializing grafana will cause whole topology not available at all?

Reply:
I've change it and make it log a Warning message